### ISSUE TITLE
Amplify console build error: Packaging lambda failed function failed. Could not find npm executable in the PATH.

### ISSUE THREAD

**Issue by Kirnos** [aws-amplify/amplify-console#1499]
**Describe the bug**
I have configured hosting for my app in AWS Amplify Console, but backend build is failing. Don't see such issues when developing the app, everything works as expected when I push the changes from my local machine to the cloud via amplify cli (i.e. amplify push).

Error message: Packaging lambda failed function failed. Could not find npm executable in the PATH.

**To Reproduce**
In my setup it happens on any attempt to trigger the build in Amplify Console

**Expected behavior**
The build succeeds and the app gets deployed to the hosting environment.

**Screenshots**
Here is the screenshot of the log with the error message:
<img width="1082" alt="Screen Shot 2021-01-25 at 11 20 27 AM" src="https://user-images.githubusercontent.com/15732276/105733759-e2392780-5eff-11eb-9449-8e9869656f9f.png">

**What is Configured?**
My app is utilizing multiple resources and several Lambda functions among them. 
One of the Lambda functions (S3UploadTrigger) is using Sharp library to resize and convert images on upload. When updating this function I always have to build it locally inside docker and manually upload the build to Lambda console. I am suspecting that this function might be causing this issue but it is not obvious.

Is there any way to find out which executable is missing in the PATH and how to set it?

**Comment by sammartinez** [comment]
This seems to be an issue with Amplify Console. I have transferred this issue to the Amplify Console repository.

**Comment by abhi7cr** [comment]
Hello @Kirnos,

It looks likes its unable to find `npm` in the PATH. You can run `which npm` to see where its installed. If you are using our default build image, we install node 8, 10 and 12 and default to node 12 and use `nvm` as our version manager. `npm` should be installed in the build image via the node installations. Looks like the lambda function expects it to be at a specific path.

**Comment by Kirnos** [comment]
Hello @abhi7cr,

Thanks for your response. I am using the default build image.
I added `which npm` to the script and it prints the following path `/root/.nvm/versions/node/v12.19.0/bin/npm`.

I have also inspected all my Lambda functions and see that each one has the file called `amplify.state` with the following content:
`{
  "pluginId": "amplify-nodejs-function-runtime-provider",
  "functionRuntime": "nodejs",
  "useLegacyBuild": true,
  "defaultEditorFile": "src/index.js"
}`
I can't find any documentation describing this file. What is the `useLegacyBuild`parameter used for?
It seems that the issue is coming from the `amplify-nodejs-function-runtime-provider` module `legacyBuild.ts` file, here is the full path from the logs pointing to this plugin `/root/.nvm/versions/node/v12.19.0/lib/node_modules/@aws-amplify/cli/node_modules/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts`

Also, I noted that some of my Lambdas use 10.x nodejs and some 12.x runtime. Should I modify all of them to have the same runtime?

**Comment by abhi7cr** [comment]
@Kirnos Yes, can you modify the lambdas so that they all use 12.x ? I’m not sure about the `useLegacyBuild` property, I can check with my colleagues and get back to you.

**Comment by Kirnos** [comment]
@abhi7cr - I updated all Lambdas to use 12.x but it didn't fix the issue.
The only function in my backend that still requires 10.x is the autogenerated userPoolGroups lambda which is defined inside .../amplify/backend/auth/userPoolGroups/template.json and always defaults to 10.x even after I update the template.

I looked at the source code of the amplify function throwing the error and see that yarn is also supported there. https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts#L44

So I ran yarn package manager for each Lambda and now I have yarn.lock files in lambdas' src folders. The issue is still the same - it can't find npm executable in the PATH.

I am stuck at this point and as a next step I am going to create another environment off the current one and try removing lambdas/resources one by one to see which one is causing the issue.

**Comment by igorimgn** [comment]
Having the same error after adding 2 new lambdas via cli

**Comment by MaximilianSchon** [comment]
I am having the same issue after fetching the project to a new computer.

**Comment by abhi7cr** [comment]
could @igorimgn  and @Kirnos  share both of you app ids and region ? I will take a look at the logs on our side.

**Comment by zfanta** [comment]
Function `runPackageManager` in `amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts` uses `execa.sync` with `cwd` option[(code)](https://github.com/aws-amplify/amplify-cli/blob/v4.41.2/packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts#L47). If `cwd` is not a valid directory path, `execa.sync` throws error.

If you add `console.log(cwd)` in `runPackageManager` then you may found invalid `cwd`. That invalid `cwd` is originated from amplify configuration files(json, yml). In my case invalid lambda function name was in `amplify/backend/amplify-meta.json`

**Comment by ttolvane** [comment]
Seeing the same issue. Locally building and pushing works fine, but cloud builds fail for the same reason: 

`Error: Packaging lambda failed function failed. Could not find npm executable in the PATH.`

**Comment by abhi7cr** [comment]
@ttolvane, could you share your appId and region ?

**Comment by ttolvane** [comment]
@abhi7cr Thanks for looking into this. I did file a support request thru AWS support with the details.

For me this got a bit annoying since I happened to deploy my backend to production using amplify cli due to model changes I needed to migrate in using couple of separate deployments. All that went as planned until I wanted to do the final push via Amplify console to get the front-end deployed due to braking changes. To my surprise this specific backend build issue then caused that to fail rendering the service unusable. To get around it quickly I just bluntly commented out amplifyPush from amplify.yml to get past the backend deployment part to the front-end deployed via Amplify console. Perhaps there is a better way to achieve the same, but it worked for the time critical deployment task at hand.

**Comment by AlessandroAnnini** [comment]
Any news about this bug?

**Comment by ludovicoserrani-nautes** [comment]
> Function `runPackageManager` in `amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts` uses `execa.sync` with `cwd` option[(code)](https://github.com/aws-amplify/amplify-cli/blob/v4.41.2/packages/amplify-nodejs-function-runtime-provider/src/utils/legacyBuild.ts#L47). If `cwd` is not a valid directory path, `execa.sync` throws error.
> 
> If you add `console.log(cwd)` in `runPackageManager` then you may found invalid `cwd`. That invalid `cwd` is originated from amplify configuration files(json, yml). In my case invalid lambda function name was in `amplify/backend/amplify-meta.json`

Thank you @zfanta, this solved for me.

The problem was that I created a function with the name in kebab case (e.g.: `my-lambda-function`) using `amplify add function`. When I pushed I ran `amplify push` from my PC then I got an error saying that the name was not valid. I then removed the function with the command `amplify remove function` and created a new one with a valid name.

Apparently, the CLI did not remove all references to the removed function. In fact, I had to remove manually the references from `backend-config.json` and from `team-provider-info.json` and then the build was successfull.

If it can be helpul the CLI version is 4.41.2.

**Comment by igorimgn** [comment]
I think my problem was also kebab naming. But there is also the issue that multiple changes to the same resource are not yet 100% smooth - there are undeleted references.

For example, when I add a lambda and then deploy it. If I change my mind and remove it later all still good. But if I add it again then thing get broken.

I did the lazy approach. Reset to a previous commit and then add the lambda with a different name and without kebab case. So probably there is some dead reference in the code.

**Comment by mgcox** [comment]
Just came across this thread... project is kebab case..... I'm really hoping this isn't the death of my project... but I'm having the same problem

**Comment by mgcox** [comment]
Resolved same errors in my project. 

Issue was newly created functions were not added to git repo.

**Comment by fkunecke** [comment]
Had this same issue because I deleted a function and amplify didn't remove all the resources correctly. I fixed it by removing the remaining references from `team-provider-info.json` and `backend-config.json`.

You may run into this if you've got a command window open in a lambda src directory. This has happened to me multiple times now because of this.

**Comment by AlessandroVol23** [comment]
For me it helped by removing a function I didn't use. No idea why but it resolved the issue.

**Comment by StackTrac3** [comment]
Empty folder and function in the backend config and team provider info caused this for me

**Comment by StackTrac3** [comment]
Empty folder and function in the backend config and team provider info caused this for me.

Had to edit several generated files - this is an annoying one.

**Comment by valtrem** [comment]
I got this error after manually renaming a Function throughout the code. (Find and replace)

Running ```amplify status``` would still show the function name with the old name.
Running ```amplify init``` fixed the discrepancy, but I would still get the error.

I finally noticed I had forgotten to rename the function's contain folder !

Renamed that too and problem fixed.

**Comment by stellarjaysoftware** [comment]
This happened to me when jumping branches during development.

- On one branch I had created some amplify resources: storage, api, and function.  They were synced to the cloud.
- I committed changes to branch #1, checked out master, and then created another branch #2.
- I then started to work on some new amplify resources in branch #2.

when trying to push new resources: "Packaging lambda failed function failed. Could not find npm executable in the PATH"

I believe since some of the amplify generated state files are not tracked by git they got out of sync.

To fix it I removed my changes in the new branch (backed them up locally), switched back to the first branch, and reverted some changes to the non-tracked amplify state files: amplify/backend/amplify-meta.json, amplify/team-provider-info.json

I'm not exactly sure what's the right way to move forward, but I thought I'd post a solution that at least got me back to a working state in case it's helpful to anyone.

**Comment by cunneen** [comment]
It would be really useful to have the `scriptName` and `CWD` as part of the error message. This is driving me nuts. I don't have any unused references in my `backend-config.json` or `team-provider-info.json`.
